src/cmd/run.go

```diff
@@ -299,48 +287,82 @@ func runCommand(container string,
 
 	envOptions := utils.GetEnvOptionsForPreservedVariables()
 
-	execArgs := constructExecArgs(container, command, detachKeysSupported, envOptions, workingDirectory)
+	runFallbackCommandsIndex := 0
+	runFallbackWorkDirsIndex := 0
+	workDir := workingDirectory
 
-	if emitEscapeSequence {
-		fmt.Printf("\033]777;container;push;%s;toolbox;%s\033\\", container, currentUser.Uid)
-	}
+	for {
+		execArgs := constructExecArgs(container, command, detachKeysSupported, envOptions, workDir)
 
-	logrus.Debugf("Running in container %s:", container)
-	logrus.Debug("podman")
-	for _, arg := range execArgs {
-		logrus.Debugf("%s", arg)
-	}
+		if emitEscapeSequence {
+			fmt.Printf("\033]777;container;push;%s;toolbox;%s\033\\", container, currentUser.Uid)
+		}
 
-	exitCode, err := shell.RunWithExitCode("podman", os.Stdin, os.Stdout, nil, execArgs...)
+		logrus.Debugf("Running in container %s:", container)
+		logrus.Debug("podman")
+		for _, arg := range execArgs {
+			logrus.Debugf("%s", arg)
+		}
 
-	if emitEscapeSequence {
-		fmt.Printf("\033]777;container;pop;;;%s\033\\", currentUser.Uid)
-	}
+		exitCode, err := shell.RunWithExitCode("podman", os.Stdin, os.Stdout, nil, execArgs...)
 
-	switch exitCode {
-	case 0:
-		if err != nil {
-			panic("unexpected error: 'podman exec' finished successfully")
-		}
-	case 125:
-		err = fmt.Errorf("failed to invoke 'podman exec' in container %s", container)
-	case 126:
-		err = fmt.Errorf("failed to invoke command %s in container %s", command[0], container)
-	case 127:
-		if pathPresent, _ := isPathPresent(container, workingDirectory); !pathPresent {
-			err = fmt.Errorf("directory %s not found in container %s", workingDirectory, container)
-		} else {
-			err = fmt.Errorf("command %s not found in container %s", command[0], container)
+		if emitEscapeSequence {
+			fmt.Printf("\033]777;container;pop;;;%s\033\\", currentUser.Uid)
 		}
-	default:
-		err = nil
-	}
 
-	if err != nil {
-		return err
+		switch exitCode {
+		case 0:
+			if err != nil {
+				panic("unexpected error: 'podman exec' finished successfully")
+			}
+			return nil
+		case 125:
+			err = fmt.Errorf("failed to invoke 'podman exec' in container %s", container)
+			return err
+		case 126:
+			err = fmt.Errorf("failed to invoke command %s in container %s", command[0], container)
+			return err
+		case 127:
+			if pathPresent, _ := isPathPresent(container, workDir); !pathPresent {
+				if runFallbackWorkDirsIndex < len(runFallbackWorkDirs) {
+					fmt.Fprintf(os.Stderr,
+						"Error: directory %s not found in container %s\n",
+						workDir,
+						container)
+
+					workDir = runFallbackWorkDirs[runFallbackWorkDirsIndex]
+					if workDir == "" {
+						workDir = currentUser.HomeDir
+					}
+
+					fmt.Fprintf(os.Stderr, "Using %s instead.\n", workDir)
+					runFallbackWorkDirsIndex++
+				} else {
+					err = fmt.Errorf("directory %s not found in container %s", workDir, container)
+					return err
+				}
+			} else {
```

Comment:
We can't assume that an exit code of 127 from `podman exec` means that either the working directory or the command was absent. The given shell might also return with 127, which will get forwarded by `podman exec`, if the last command that was attempted inside the shell was absent. In such cases this will create a loop.

See https://github.com/containers/toolbox/pull/872